### PR TITLE
[Dubbo-2064] Ipv6 support

### PR DESCRIPTION
**Pull request by ralf0131**
## What is the purpose of the change

Add ipv6 support for Dubbo #2064 

## Brief changelog

* Support parsing ipv6 address in url
* Support ipv6 multicast network
* ipv6 support is available if `java.net.preferIPv6Addresses` is set to `true`

## Verifying this change

How to test ipv6 support:

dubbo-demo-provider.xml

```xml
<beans xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
       xmlns:dubbo="http://dubbo.apache.org/schema/dubbo"
       xmlns="http://www.springframework.org/schema/beans"
       xsi:schemaLocation="http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans-4.3.xsd
       http://dubbo.apache.org/schema/dubbo http://dubbo.apache.org/schema/dubbo/dubbo.xsd">

    <!-- provider's application name, used for tracing dependency relationship -->
    <dubbo:application name="demo-provider"/>

    <!-- use multicast registry center to export service -->
    <dubbo:registry address="multicast://FF02:0:0:0:0:0:0:FC:1234"/>

    <!-- use dubbo protocol to export service on port 20880 -->
    <dubbo:protocol name="dubbo" port="20880"/>

    <!-- service implementation, as same as regular local bean -->
    <bean id="demoService" class="org.apache.dubbo.demo.provider.DemoServiceImpl"/>

    <!-- declare the service interface to be exported -->
    <dubbo:service interface="org.apache.dubbo.demo.DemoService" ref="demoService"/>

</beans>
```

dubbo-demo-consumer.xml

```xml
<beans xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
       xmlns:dubbo="http://dubbo.apache.org/schema/dubbo"
       xmlns="http://www.springframework.org/schema/beans"
       xsi:schemaLocation="http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans-4.3.xsd
       http://dubbo.apache.org/schema/dubbo http://dubbo.apache.org/schema/dubbo/dubbo.xsd">

    <!-- consumer's application name, used for tracing dependency relationship (not a matching criterion),
    don't set it same as provider -->
    <dubbo:application name="demo-consumer"/>

    <!-- use multicast registry center to discover service -->
    <dubbo:registry address="multicast://FF02:0:0:0:0:0:0:FC:1234"/>

    <!-- generate proxy for the remote service, then demoService can be used in the same way as the
    local regular interface -->
    <dubbo:reference id="demoService" check="false" interface="org.apache.dubbo.demo.DemoService"/>

</beans>
```

Add `-Djava.net.preferIPv6Addresses=true` before starting Provider/Consumer.

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/incubator-dubbo/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [x] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/incubator-dubbo/tree/master/dubbo-test).
- [x] Run `mvn clean install -DskipTests` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/incubator-dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Comment by luyuanwan** [comment]
Why not create a IPV6Utils class? @ralf0131   :)

**Comment by ralf0131** [comment]
@luyuanwan  Good suggestion!

For me, I think the current refactoring is not large enough to make that change. You can always improve it as you wish, and send the pull request.

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/NetUtils.java [hunk 1]
```diff
@@ -194,26 +194,22 @@ private static InetAddress getLocalAddress0() {
...
-                            while (addresses.hasMoreElements()) {
-                                try {
-                                    InetAddress address = addresses.nextElement();
-                                    if (isValidAddress(address)) {
-                                        return address;
-                                    }
-                                } catch (Throwable e) {
-                                    logger.warn(e);
-                                }
+            while (interfaces.hasMoreElements()) {
```
Comment (zonghaishang) [inline dubbo-common/src/main/java/org/apache/dubbo/common/utils/NetUtils.java]:
```
    public static Enumeration<NetworkInterface> getNetworkInterfaces()
        throws SocketException {
        final NetworkInterface[] netifs = getAll();

        // specified to return null if no network interfaces
        if (netifs == null)
            return null;
```

Should check the interfaces property  null value

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/NetUtils.java [hunk 2]
```diff
@@ -194,26 +194,22 @@ private static InetAddress getLocalAddress0() {
...
+                            if (isValidAddress(address)) {
+                                return address;
                             }
+                        } catch (Throwable e) {
+                            logger.warn(e);
                         }
-                    } catch (Throwable e) {
-                        logger.warn(e);
                     }
+                } catch (Throwable e) {
```
Comment (zonghaishang) [inline dubbo-common/src/main/java/org/apache/dubbo/common/utils/NetUtils.java]: rename e to ignored will be better

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/NetUtils.java [hunk 3]
```diff
@@ -194,26 +194,22 @@ private static InetAddress getLocalAddress0() {
...
+                try {
+                    NetworkInterface network = interfaces.nextElement();
+                    Enumeration<InetAddress> addresses = network.getInetAddresses();
+                    while (addresses.hasMoreElements()) {
+                        try {
+                            InetAddress address = addresses.nextElement();
+                            if (isValidAddress(address)) {
+                                return address;
                             }
+                        } catch (Throwable e) {
```
Comment (zonghaishang) [inline dubbo-common/src/main/java/org/apache/dubbo/common/utils/NetUtils.java]: rename e to ignored will be better

File: dubbo-common/src/test/java/org/apache/dubbo/common/utils/NetUtilsTest.java [hunk 4]
```diff
@@ -180,4 +181,23 @@ public void testToURL() throws Exception {
         String url = NetUtils.toURL("dubbo", "host", 1234, "foo");
         assertThat(url, equalTo("dubbo://host:1234/foo"));
     }
+
+    @Test
+    public void testIsValidV6Address() {
+        System.setProperty("java.net.preferIPv6Addresses", "true");
```
Comment (zonghaishang) [inline dubbo-common/src/test/java/org/apache/dubbo/common/utils/NetUtilsTest.java]:
```
String saved = System.setProperty("java.net.preferIPv6Addresses");
System.setProperty("java.net.preferIPv6Addresses", "true");
...
System.setProperty("java.net.preferIPv6Addresses", saved);
```
This way is more elegant and will not break subsequent tests.

Comment (yangjinhe) [reply dubbo-common/src/test/java/org/apache/dubbo/common/utils/NetUtilsTest.java]: String saved = System.getProperty("java.net.preferIPv6Addresses");
System.setProperty("java.net.preferIPv6Addresses", "true");
...
System.setProperty("java.net.preferIPv6Addresses", saved);

Comment (ralf0131) [reply dubbo-common/src/test/java/org/apache/dubbo/common/utils/NetUtilsTest.java]: Good catch, it is better.

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/NetUtils.java (hunk 1 above)
Comment (ralf0131) [reply dubbo-common/src/main/java/org/apache/dubbo/common/utils/NetUtils.java]: Thanks, I am using Java 10, the implementation is different with Java 8. In Java 10 the implementation guarantees no null value is returned. I will add the null check.

**Comment by ralf0131** [comment]
@zonghaishang I've improved the code based on your review comments, could you take a look?

**Comment by zonghaishang** [comment]
It looks good to me.

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/NetUtils.java [hunk 5]
```diff
@@ -194,26 +194,22 @@ private static InetAddress getLocalAddress0() {
...
+                            logger.warn(e);
                         }
-                    } catch (Throwable e) {
-                        logger.warn(e);
                     }
+                } catch (Throwable e) {
+                    logger.warn(e);
```
Comment (dirkzhou222) [inline dubbo-common/src/main/java/org/apache/dubbo/common/utils/NetUtils.java]: ddddd

**Pull request merged**